pkg/util/fileutil/filesystem.go

```diff
@@ -29,37 +28,101 @@ var (
 	targetPathRe = regexp.MustCompile(`[\\|\/]+pods[\\|\/]+(.+?)[\\|\/]+volumes[\\|\/]+kubernetes.io~csi[\\|\/]+(.+?)[\\|\/]+mount$`)
 )
 
-// GetMountedFiles returns all the mounted files names with filepath base as key
+// GetMountedFiles returns all the mounted files mapping their path relative to
+// targetPath to the absolute paths.
+//
+// This will filter out files by atomic_writer (which reserves file prefixed
+// `..` and follows the symlinks created by atomic_writer).
 func GetMountedFiles(targetPath string) (map[string]string, error) {
 	paths := make(map[string]string)
-	// loop thru all the mounted files
-	var files []string
-	err := filepath.Walk(targetPath, func(path string, fi os.FileInfo, err error) error {
-		if err != nil {
-			return err
-		}
-		if fi.IsDir() {
-			return nil
+
+	// atomic writer will write data to, but filepath.Walk does not follow
+	// symlinks
+	d, err := os.ReadDir(targetPath)
+	if err != nil {
+		return paths, err
+	}
+
+	// for each item in the targetPath, walk that item
+	for _, entry := range d {
+		// skip the reserved paths of targetPath/..*
+		if strings.HasPrefix(entry.Name(), "..") {
+			continue
 		}
-		relpath, err := filepath.Rel(targetPath, path)
+
+		info, err := entry.Info()
 		if err != nil {
-			return err
+			continue
 		}
-		files = append(files, relpath)
-		return nil
-	})
-	if err != nil {
-		return nil, fmt.Errorf("failed to list all files in target path %s, err: %v", targetPath, err)
-	}
 
-	sep := "/"
-	if strings.HasPrefix(targetPath, "c:\\") {
-		sep = "\\"
-	} else if strings.HasPrefix(targetPath, `c:\`) {
-		sep = `\`
-	}
-	for _, file := range files {
-		paths[file] = targetPath + sep + file
+		// the path to the file relative to targetPath
+		// i.e. foo
+		base := info.Name()
+
+		// the path before following the symlink
+		// i.e. targetPath/foo
+		p := filepath.Join(targetPath, info.Name())
+
+		// for symlinks in the targetPath...
+		if info.Mode()&os.ModeSymlink != 0 {
+			// the resolved relative path
+			// i.e. ..data/foo
+			actual, err := os.Readlink(p)
+			if err != nil {
+				continue
+			}
+
+			// the root path to walk
+			// i.e. targetPath/..data/foo
+			root := filepath.Join(targetPath, actual)
+
+			if err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+				// if there was an error walking path immediately propagate it
```

Comment:
The majority of `filepath.Walk` in the if..else loop is similar except for the paths. Could we extract this into a common function?

Reply:
After sleeping on this I've refactored a few times to reduce duplication and hopefully it reads better now